pytorch_lightning/trainer/connectors/accelerator_connector.py

```diff
@@ -73,959 +71,762 @@
-                return env_type()
 
-        return LightningEnvironment()
-
-    def set_distributed_mode(self, strategy: Optional[str] = None):
+    def _lazy_init_strategy(self):
+        # set strategy properties
+        self.strategy.accelerator = self.accelerator
+        if self.precision_plugin:
+            self.strategy.precision_plugin = self.precision_plugin
+        if self.checkpoint_io:
+            self.strategy.checkpoint_io = self.checkpoint_io
+        if hasattr(self.strategy, "cluster_environment"):
+            self.strategy.cluster_environment = self.cluster_environment
+        if hasattr(self.strategy, "parallel_devices"):
+            if self.strategy.parallel_devices:
+                self._parallel_devices = self.strategy.parallel_devices
+            else:
+                self.strategy.parallel_devices = self._parallel_devices
+        if hasattr(self.strategy, "num_nodes"):
+            self.strategy._num_nodes = self._num_nodes_flag
+        if hasattr(self.strategy, "sync_batchnorm"):
+            self.strategy.sync_batchnorm = self.sync_batchnorm
+        if hasattr(self.strategy, "set_world_ranks"):
+            self.strategy.set_world_ranks()
 
-        if strategy is None and self.is_training_type_in_plugins:
-            return
+        from pytorch_lightning.utilities import _IS_INTERACTIVE
 
-        if strategy is not None and strategy in StrategyRegistry:
-            self.distributed_backend = StrategyRegistry[strategy]["distributed_backend"]
-        elif strategy is not None:
-            self.distributed_backend = strategy
+        interactive_compatible_strategy = ("dp", "ddp_spawn", "ddp_sharded_spawn", "tpu_spawn")
+        if _IS_INTERACTIVE and self.strategy.distributed_backend not in interactive_compatible_strategy:
+            raise MisconfigurationException(
+                f"`Trainer(strategy={self.strategy.distributed_backend!r})` or"
+                f" `Trainer(accelerator={self.strategy.distributed_backend!r})` is not compatible with an interactive"
+                " environment. Run your code as a script, or choose one of the compatible backends:"
+                f" {', '.join(interactive_compatible_strategy)}."
+                " In case you are spawning processes yourself, make sure to include the Trainer"
+                " creation inside the worker function."
+            )
 
-        if isinstance(self.distributed_backend, Accelerator):
-            return
+        # TODO should be moved to _strategy_check_and_fallbacks().
+        # Current test check precision first, so keep this check here to meet error order
+        if isinstance(self.accelerator, TPUAccelerator) and not isinstance(
+            self.strategy, (SingleTPUStrategy, TPUSpawnStrategy)
+        ):
+            raise ValueError(
+                "The `TPUAccelerator` can only be used with a `SingleTPUStrategy` or `TPUSpawnStrategy`,"
+                f" found {self.strategy}."
+            )
```

Comment:
i don't get this check. what if someone implements a custom strategy which uses the TPU accelerator? why do we need to assert this?

Reply:
Yeah agree. This is for backward compatibility. How about we revisit all the Misconfig and ValueErrors after this PR? If we modify logic, this PR will be huge with all the tests updates

Reply:
#11449 follow up item 6